### PR TITLE
prov/efa: Adding some checking around enabled shared memory transfers

### PR DESCRIPTION
**Pull request by richwelchaws**
There are currently a couple of places where a global variable is used
to control whether or not shared memory is used for intranode
communnication. This can be dangerous if for some reason this gets
out of sync with the actual underlying structures.

This patch changes things to that we primarily check the underlying
structures, and only set those after checking the global variable
once. There are some occurances where making this change isn't
possible but for the most part this patch makese these checks more
defensive.

Also, in some places the check on the global variable is redundant
and any inconsistency would represent an error state, so these checks
were changes to asserts.

Signed-off-by: Rich Welch <rlwelch@amazon.com>

### CODE REVIEW

File: man/fi_shm.7.md [hunk 1]
```diff
@@ -25,7 +25,7 @@ offers the following support:
 : The provider supports only endpoint type *FI_EP_RDM*.
 
 *Endpoint capabilities*
-: Endpoints cna support any combinations of the following data transfer
+: Endpoints can support any combinations of the following data transfer
```
Comment (rwespetal) [inline man/fi_shm.7.md:28]: Please move the shm changes to a separate commit and/or PR

File: prov/efa/src/efa_fabric.c [hunk 2]
```diff
@@ -1006,6 +1006,8 @@ int efa_fabric(struct fi_fabric_attr *attr, struct fid_fabric **fabric_fid,
 				    &efa_fabric->shm_fabric, context);
 		if (ret)
 			goto err_close_util_fabric;
+	} else {
+		efa_fabric->shm_fabric=NULL;
```
Comment (rwespetal) [inline prov/efa/src/efa_fabric.c:1010]: Space between `=` is needed

**Review by rwespetal — COMMENTED** [review]
I'm seeing at least one spot where we can avoid using the environment variable - `efa_av_open` is one. Should be able to check whether shm_fabric is initialized there.

File: man/fi_shm.7.md (hunk 1 above)
Comment (richwelchaws) [reply man/fi_shm.7.md]: Ah, I see these are for another provider. I will remove them from this PR and handle outside of this change.

File: prov/efa/src/efa_fabric.c (hunk 2 above)
Comment (richwelchaws) [reply prov/efa/src/efa_fabric.c]: Done, thanks!

**Comment by richwelchaws** [comment]
> I'm seeing at least one spot where we can avoid using the environment variable - `efa_av_open` is one. Should be able to check whether shm_fabric is initialized there.

Good catch - I've resolved the instance in `efa_av_open` (please see update). Thanks!

File: prov/efa/src/efa_av.c [hunk 3]
```diff
@@ -840,8 +840,8 @@ static int efa_av_close(struct fid *fid)
 	}
 
 	if (av->ep_type == FI_EP_RDM) {
-		if (rxr_env.enable_shm_transfer && av->shm_rdm_av &&
-		    &av->shm_rdm_av->fid) {
+		if (av->shm_rdm_av && &av->shm_rdm_av->fid) {
```
Comment (rwespetal) [inline prov/efa/src/efa_av.c:843]: Sorry, missed this earlier, isn't `&av->shm_rdm_av->fid` always going to be true based on the struct def? I think that can be dropped/wonder why this was there in the first place.

**Review by rwespetal — APPROVED** [review]

File: prov/efa/src/efa_av.c (hunk 3 above)
Comment (richwelchaws) [reply prov/efa/src/efa_av.c]: You are totally right. Not sure of the origins of it but I've cleaned it up in a new push. Good catch!

File: prov/efa/src/rxr/rxr_domain.c [hunk 4]
```diff
@@ -190,10 +190,9 @@ int rxr_domain_open(struct fid_fabric *fabric, struct fi_info *info,
 				  	util_domain.domain_fid);
 
 	/* Open shm provider's access domain */
-	if (rxr_env.enable_shm_transfer) {
-		efa_fabric = container_of(fabric, struct efa_fabric,
-					util_fabric.fabric_fid);
-
+	efa_fabric = container_of(fabric, struct efa_fabric,
+				util_fabric.fabric_fid);
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_domain.c:194]: indentation off here. Also because the line shifted to the left, so there is no need to be two line.

Comment (richwelchaws) [reply prov/efa/src/rxr/rxr_domain.c]: Fixed! Thanks.

**Review by wzamazon — APPROVED** [review]

**Comment by richwelchaws** [comment]
bot:aws:retest
